comp-bot/src/main/java/frc/robot/odometry/CustomOdometry.java

```diff
@@ -0,0 +1,144 @@
+package frc.robot.odometry;
+
+import dev.doglog.DogLog;
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.geometry.Transform2d;
+import edu.wpi.first.math.geometry.Translation2d;
+import edu.wpi.first.math.kinematics.SwerveDriveKinematics;
+import edu.wpi.first.math.kinematics.SwerveDriveOdometry;
+import edu.wpi.first.math.kinematics.SwerveModulePosition;
+
+public class CustomOdometry extends SwerveDriveOdometry {
+  private final int numberOfModules;
+  private final Translation2d[] robotRelativeModuleOffsets;
+
+  private Pose2d previousRobotPose = new Pose2d();
+  private final SwerveModulePosition[] previousWheelPositions;
+
+  public CustomOdometry(
+      int numberOfModules,
+      SwerveDriveKinematics kinematics,
+      Rotation2d gyroAngle,
+      SwerveModulePosition[] modulePositions) {
+    super(kinematics, gyroAngle, modulePositions);
+    this.numberOfModules = numberOfModules;
+    robotRelativeModuleOffsets = new Translation2d[numberOfModules];
+    previousWheelPositions = new SwerveModulePosition[numberOfModules];
+
+    for (int i = 0; i < numberOfModules; i++) {
+      robotRelativeModuleOffsets[i] =
+          new Translation2d(kinematics.getModules()[i].getX(), kinematics.getModules()[i].getY());
+      previousWheelPositions[i] = new SwerveModulePosition();
+    }
+  }
+
+  private static Translation2d getModuleDisplacement(
+      double previousAngleRadians,
+      double previousDistanceMeters,
+      double currentAngleRadians,
+      double currentDistanceMeters) {
+    // First, calculate difference between previous and current angles and distances
+    double angleDifferenceRadians = currentAngleRadians - previousAngleRadians;
+    double arcLength = currentDistanceMeters - previousDistanceMeters;
+
+    // *If angle difference is 0 then we can just use a straight line instead of an arc
+    if (angleDifferenceRadians == 0) {
+      return new Translation2d(arcLength, new Rotation2d(currentAngleRadians));
```

Comment:
this is creating a new `Rotation2d` object, which is unnecessary since the function could be called with `Rotation2d` objects as parameters (instead of `double`s). we can reduce the GC pressure this causes by avoiding new instantiations

which, actually, why not just have this function take 2 `SwerveModulePosition` objects? we don't need to split it out into 4 different parameters